### ISSUE TITLE
Adding tracking information via API

### ISSUE THREAD

**Issue by xiaoleih41** [paypal/PayPal-REST-API-issues#158]
*Migrated issue from https://github.com/paypal/PayPal-PHP-SDK/issues/951
Known work around: Log into PayPal account and enter tracking to the transaction detail. 
For internal bookkeeping: PPPLPAYPT-28860 
Need API option to [add delivery tracking for transactions](https://www.paypal.com/uk/selfhelp/article/how-to-add-tracking-information-for-a-payment-you-received.-faq3230).

**Comment by sanmai** [comment]
Here's [what Keegan from the Shopify Guru team found after talking with PayPal](https://ecommerce.shopify.com/c/payments-shipping-fulfilment/t/putting-tracking-number-on-paypal-336098):

> ...They advised that the "Tracking number box" on your merchant dashboard is usually only used if you're using the Paypal Shipping Label area.
> 
> https://www.paypal.com/ca/webapps/mpp/shipping-centre
> 
> Meaning, general use for outside providers (like Shopify) is to not hand off shipping information, since it's all being processed our side (order confirmation/notification) instead of on Paypal. Paypal Express Checkout just being the money acceptance, not the tracking/fulfillment/shipping for (most) businesses.

This could explain why there is no API to add a tracking information, and will not be.

**Comment by hamzaydia** [comment]
PayPal needs to add a clarification about "Add tracking information" that it's only for "Paypal Shipping Label" to avoid the confusion.

**Comment by vincevanmeer** [comment]
But on PayPal's site it literally says that some payments may not be eligible for sellers protection if no tracking information is set/provided — So I definitely see use for this feature.

**Comment by sanmai** [comment]
Right, that's what they'll ask anyway to provide once there is a claim filed. No real reason to bother about it before that.

**Comment by speller** [comment]
21st century is going, Earth people plan to colony Mars, but PayPal still have no API to add tracking API (they even don't know there are users except Ebay). Poorest IT team I ever met in so globally known company. We are processing hundred parcels every day paid by PayPal (and NO, we don't use Ebay for that) and we have to spread money for staff who manually enter tracking numbers and make errors. Every technician who support PayPal from seller's side dream to forget PayPal once forever like a nightmare. 21st century, the century of IT and automation, by PayPal users still obliged to use hand work.

**Comment by speller** [comment]
Could you describe more about PayPal template please? As I googled they don't provide API for labels too. So how to avoid hand work in this case? We need to automate the whole process and avoid maximum of human labor.

**Comment by speller** [comment]
https://www.paypal.com/dm/webapps/mpp/security/seller-protection

> You must provide online tracking to be eligible for protection

**Comment by sanmai** [comment]
Yeah, as far as I know that's not a problem to provide that *after* a claim arises. But you should definitely reconfirm that with your manager.

**Comment by speller** [comment]
Sorry for my expression we just have a lot of pain because of absence of tracking upload API...

**Comment by sanmai** [comment]
I feel your pain, but PayPal has nothing to do with that. What you should really do is not spend time and money hand-inputting tracking number, but get in touch with your customers about shipments, ask them to reply to the notification email if there is just about any questions (no `no-reply@` bullshit!). And respond to all questions, offer a hotline, et cetera. All is in your hands.

**Comment by speller** [comment]
Answering to customers is also takes time... ) 
So as I mentioned above - PayPal sellers have to feel pain and to keep human labor for primitive operations which can be automated very easily.

**Comment by sanmai** [comment]
Isn't it better to answer a friendly customer than to receive a straightly unpleasant claim? I can't see why one would choose second over the first, but if that's the case, I have nothing to say.

**Comment by speller** [comment]
We are also sending notifications to customers but as human staff closes transactions manually and make mistakes (incorrect tracking number or forgotten to close) customers are regularly claims PayPal on non-received item instead of contacting us first. This is the real-world customers, yeah. Some of them ask us first, but some of them claims PayPal at the same time when contacting us and the rest of them claims without contacting us. That's a particular kind of human behaviour. We send our parcels all around the world so we have the wide range of cultures and mindsets among our customers.

**Comment by sanmai** [comment]
Well, I have the opposite experience... If you really want to automate that, there's Selenium WebDriver, you can do whatever there is you desire using it. Fill all kinds of web forms, etc.

**Comment by speller** [comment]
That's the possible solution, until PayPal decide to use anti-robot techiques ) My friend from another company who use Selenium for PayPal automation shown me today the Google's "I'm not a robot" captcha captured by the robot. And that's why I've made new invetigation on is something changed since last time I was looking for tracking API and I googled this topic.

**Comment by hamzaydia** [comment]
Maybe PayPal didn't wants to spend money on storing useless data, Because when we talk an about API it means all the Shopify, WooCommerce... stores will be linked automatically, And because just a small percentage of the transactions that being disputed It is waste of resources to store all the world shipping tracking numbers and status, That's my guess.

**Comment by speller** [comment]
From the technical point of view storing the tracking number in a DB table costs $0. Or maybe $1 per ten billion of records. So it can not be a considerable reason.

**Comment by AndrewPenry** [comment]
A PayPal Customer service rep told us over the phone that it would be helpful for the dispute team if merchants added the tracking numbers to the transactions BEFORE disputes happen. If the customer service team is going to tell us to do this, then I think that it is a reasonable request that there be a way to automate it via the API.

**Comment by Arc108** [comment]
There now an app for this in chrome https://chrome.google.com/webstore/detail/paypal-tracking-upload/oaacedaplbcgmjkdkgmocgcdgdiggjna

**Comment by mariankayabrands** [comment]
Could you prove that is working?

**Comment by Arc108** [comment]
> I tested the above app https://chrome.google.com/webstore/detail/paypal-tracking-upload/oaacedaplbcgmjkdkgmocgcdgdiggjna yesterday. It is not working.
> Could you prove that is working?

I've been using it for a few weeks for hundreds of orders a day it works great. What problem are you having?

**Comment by mariankayabrands** [comment]
Could you provide me your CSV header?
What version of Google Chrome you are using and what is your OS?

**Comment by Arc108** [comment]
Here's my header: 
Date | Order # | tracking # code | First Name | Last Name | Address 1 | Address 2 | City | State | State 2 Letter | Country | SKU | Cost | Zip Code | Phone # | Product Name | Variant | Quantity | Email

I'm using the latest version of Chrome Version 71.0.3578.98 (Official Build) (64-bit)

And my OS is Mac Mojave Version 10.14.1 (18B75)

Did you check the Info tab on the app and follow the instructions? 

Hope this helps!

**Comment by mariankayabrands** [comment]
Could you please try to Order with Transaction and process a single line?

**Comment by mortdale** [comment]
PayPal has the API already. If you sell on eBay and upload tracking number via eBay, it will be automagically  updated in PayPal. so the API must be there already, just not open to public.

**Comment by Light2** [comment]
Adding tracking numbers to Paypal is extremely important for businesses that are starting to make serious revenue, you will get hit by a rolling reserve and a 60-90 day hold on your money if you don't input your tracking numbers. Obviously this will kill your cash flow and possibly your business.

However, entering tracking numbers manually if you're pulling 1000+ orders a day is an absolute pain. I really hope PayPal reconsiders this issue, it would help a lot of merchants that are currently forced to use 3rd party automation apps for this.

**Comment by hamzaydia** [comment]
Years... and there is no official clarification?

**Comment by sanmai** [comment]
> you will get hit by a rolling reserve and a 60-90 day hold on your money if you don't input your tracking numbers

I this an official reasoning?
